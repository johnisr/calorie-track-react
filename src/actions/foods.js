import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment';

export const addFood = food => ({
  type: 'ADD_FOOD',
  food: { id: uuid(), ...food },
});

export const startAddFood = (data = {}) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    const {
      name = '',
      amount = '0',
      unit = 'grams',
      carbohydrates = 0,
      protein = 0,
      fat = 0,
      calories = 0,
      timesUsed = 0,
      createdAt = moment(),
    } = data;
    const food = { name, amount, unit, carbohydrates, protein, fat, calories, timesUsed, createdAt };
    try {
      const res = await database.ref(`users/${uid}/foods`).push(food);
      dispatch(addFood({ id: res.key, ...food }));
    } catch (e) {

    }
  };
}

export const editFood = (id, updates) => ({
  type: 'EDIT_FOOD',
  id,
  updates,
});

export const startEditFood = (id, updates) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    try {
      await database.ref(`users/${uid}/foods/${id}`).update(updates);
      dispatch(editFood(id, updates));
    } catch (e) {
      
    }
  };
};

export const removeFood = ({ id }) => ({
  type: 'REMOVE_FOOD',
  id,
});

export const startRemoveFood = ({ id }) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    try {
      await database.ref(`users/${uid}/foods/${id}`).remove();
      dispatch(removeFood({ id }));
    } catch (e) {
      
    }
  };
};

export const setFoods = foods => ({
  type: 'SET_FOODS',
  foods,
});

export const startSetFoods = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    try {
      const snapshot = await database.ref(`users/${uid}/foods`).once('value');
      const foods = [];
      snapshot.forEach((food) => {
        foods.push({
          id: food.key,
          ...food.val(),
        });
      });
      dispatch(setFoods(foods));
    } catch (e) {

    }
  };
};