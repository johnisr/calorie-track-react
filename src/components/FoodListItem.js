import React from 'react';

const FoodListItem = ({ name, amount, unit, protein, carbohydrates, fat, calories, id }) => (
  <div className="listItem">
    <p className="listItem__text">{name}</p>
    <p className="listItem__text">{amount} {unit}</p>
    <p className="listItem__text">{carbohydrates}</p>
    <p className="listItem__text">{protein}</p>
    <p className="listItem__text">{fat}</p>
    <p className="listItem__text">{calories}</p>
  </div>
);

export default FoodListItem;
