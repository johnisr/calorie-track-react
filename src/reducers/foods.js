const foodsDefaultState = [];

export default (state = foodsDefaultState, action) => {
  switch (action.type) {
    case 'ADD_FOOD':
      return [...state, action.food];
    case 'EDIT_FOOD':
      return state.map((food) => {
        if (food.id === action.id) {
          return {
            ...food,
            ...action.updates,
          };
        }
        return food;
      });
    case 'REMOVE_FOOD':
      return state.filter(({ id }) => id !== action.id);
    case 'SET_FOODS':
      return action.foods;
    default:
      return state;
  }
}