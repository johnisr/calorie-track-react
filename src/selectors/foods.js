export default (foods, { name, sortBy, maxFoodsShown, offset }) => {
  return foods
    .filter((food) => {
      const nameMatch = food.name.toLowerCase().includes(name.toLowerCase());
      return nameMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'usage') {
        return a.timesUsed < b.timesUsed ? 1 : -1;
      } else if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
    })
    .slice((offset * maxFoodsShown), (offset * maxFoodsShown) + maxFoodsShown);
}