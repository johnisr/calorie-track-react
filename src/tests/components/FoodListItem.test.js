import React from 'react';
import { shallow } from 'enzyme';
import FoodListItem from '../../components/FoodListItem';
import foods from '../fixtures/foods';

it('should render FoodListItem correctly', () => {
  const wrapper = shallow(<FoodListItem {...foods[2]}/>);
  expect(wrapper).toMatchSnapshot();
});