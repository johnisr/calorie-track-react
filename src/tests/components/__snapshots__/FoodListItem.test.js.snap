// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render FoodListItem correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FoodListItem
    amount={24}
    calories={163}
    carbohydrates={2.5}
    createdAt={-1000}
    fat={14}
    id="3"
    name="Almonds"
    protein={6}
    timesUsed={2}
    unit="pieces"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="listItem__text"
>
          Almonds
</p>,
        <p
          className="listItem__text"
>
          24
           
          pieces
</p>,
        <p
          className="listItem__text"
>
          2.5
</p>,
        <p
          className="listItem__text"
>
          6
</p>,
        <p
          className="listItem__text"
>
          14
</p>,
        <p
          className="listItem__text"
>
          163
</p>,
      ],
      "className": "listItem",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Almonds",
          "className": "listItem__text",
        },
        "ref": null,
        "rendered": "Almonds",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            24,
            " ",
            "pieces",
          ],
          "className": "listItem__text",
        },
        "ref": null,
        "rendered": Array [
          24,
          " ",
          "pieces",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 2.5,
          "className": "listItem__text",
        },
        "ref": null,
        "rendered": 2.5,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 6,
          "className": "listItem__text",
        },
        "ref": null,
        "rendered": 6,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 14,
          "className": "listItem__text",
        },
        "ref": null,
        "rendered": 14,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 163,
          "className": "listItem__text",
        },
        "ref": null,
        "rendered": 163,
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="listItem__text"
>
            Almonds
</p>,
          <p
            className="listItem__text"
>
            24
             
            pieces
</p>,
          <p
            className="listItem__text"
>
            2.5
</p>,
          <p
            className="listItem__text"
>
            6
</p>,
          <p
            className="listItem__text"
>
            14
</p>,
          <p
            className="listItem__text"
>
            163
</p>,
        ],
        "className": "listItem",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Almonds",
            "className": "listItem__text",
          },
          "ref": null,
          "rendered": "Almonds",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              24,
              " ",
              "pieces",
            ],
            "className": "listItem__text",
          },
          "ref": null,
          "rendered": Array [
            24,
            " ",
            "pieces",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 2.5,
            "className": "listItem__text",
          },
          "ref": null,
          "rendered": 2.5,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 6,
            "className": "listItem__text",
          },
          "ref": null,
          "rendered": 6,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 14,
            "className": "listItem__text",
          },
          "ref": null,
          "rendered": 14,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 163,
            "className": "listItem__text",
          },
          "ref": null,
          "rendered": 163,
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
